### PR TITLE
HOOQ coding test (Khoon Hiang)

### PR DESCRIPTION
**Pull request by gohkhoonhiang**
Hi @diorahman ,

Here's my **first** attempt at Node.js.

Appreciate your feedback if any of my code is questionable or not adhering to standards.

Just want to provide my comment.

1. The `naive()` function has nested callbacks which make the code difficult to follow, so I decided to break out the functions and callbacks.

2. I used `async.waterfall` for the async implementation so that I can pass the argument from the previous callback to the next callback, eg. `loadedCsv` to the `parse` callback.

3. For the promise implementation, `sendSms(line).catch(logToS3).catch(printDebug);`, the logging to s3 and debug log for logging error are triggered on promise rejected. This is to keep to the logic of `logging if send sms err` and `debug if log err`.

4. For the stream implementation, I created the `readline Interface` to read the file stream line-by-line, then pipe the data to `parser` for processing. Since the data is piped to `parser` line-by-line, I'm not sure how to get the index position, so I removed the `debug` for the index.

5. I created the transform function (firstname + lastname) in `helper.js` so that all 4 implementations can call the function.

Hope to hear from you soon.

Thanks!

Khoon

### CODE REVIEW

File: parse-callback.js [hunk 1]
```diff
@@ -8,46 +8,84 @@ const helper = require('./helper');
...
-                        }
-                    });
-                }
+            sendSms(line);
+        }
 
-                index++;
-            }
-        });
+        index++;
```
Comment (diorahman) [inline parse-callback.js]: Hi!

I'm not sure you'll find it useful. But, I want to know, why did you leave this line (the `index++`) here?

File: parse-promise.js [hunk 2]
```diff
@@ -1 +1,97 @@
-// please use promise approach to fight the naive one in parse-callback.js
+'use strict';
+
+const debug = require('debug')('hello');
+
+const fs = require('fs');
+const parse = require('csv-parse');
+const helper = require('./helper');
+const Promise = require('promise');
```
Comment (diorahman) [inline parse-promise.js]: Promise is supported natively in node. Unless you're doing something fancy, I don't think you need another lib for this.

File: parse-promise.js [hunk 3]
```diff
@@ -1 +1,97 @@
...
+        // backlog #3
+        line = helper.transform(line);
+
+        if (index > 0) {
+            debug(`sending data index: ${index - 1}`);
+
+            sendSms(line).catch(logToS3).catch(printDebug);
+        }
+
+        index++;
```
Comment (diorahman) [inline parse-promise.js]: Ditto. Why do you need this?

File: parse-stream.js [hunk 4]
```diff
@@ -1,3 +1,60 @@
...
+'use strict';
+
+const debug = require('debug')('hello');
+
+const fs = require('fs');
+const readline = require('readline');
+const parse = require('csv-parse');
+const helper = require('./helper');
+
+var filename = '/sample.csv';
```
Comment (diorahman) [inline parse-stream.js]: Do you know the diffs between var, const and let?

**Comment by gohkhoonhiang** [comment]
Hi @diorahman,

Thanks for your comments.

**Question 1 & 3:**

I was too focused on solving the callbacks and didn't pay attention to such details and just copied-pasted the code when I extracted the functions. `index++` is redundant because `(let index in parsed)` will declare a new instance of `index` at the start of the loop so this increment has no use.

**Question 2:**

I'm not aware Node.js comes with promise. As I was doing this part and did a search on how to use promise, I was shown promise.js as the first result. After you mentioned Node.js has native promise, I tried searching again and read that it is a ES6 feature. Are you referring to that as native?

**Question 4:**

It was the first time I encountered `let` and `const` when doing this task and I had a quick lookup during that time. But I didn't know enough to really use them correctly, so I was doing the test to the best of my existing JS knowledge, which is limited to just `var`, and I used `let` or `const` according to the sample code you have given. Since you asked now, I decided to do a little more research into the 3 keywords. Here's my understanding after reading more:

`const` variables, once declared and initialized, cannot be redeclared or reassigned later. Attempting to do so will cause errors. However, if the variable is assigned to a non-primitive, eg. `const x = {a:1}`, it is possible to reassign the value of `x.a`.

`let` variables are limited for reference only in the block (within the nearest enclosing braces) they are declared. And there cannot be more than one `let` declaration for the same variable name within the same scope. Once out of the scope, the variable cannot be referenced. However, it is possible to declare `let` variables in the global scope, reference and redeclare them inside functions/blocks, and assign new values to them.

`var` variables are free for reference in a function scope if declared within a function, or in the global scope if declared anywhere outside a function. The difference of this with `let` is, such variables can be redeclared within the same scope, eg. `function f() { var x = 1; var x = 2; console.log(x); }` is valid.

For both `let` and `var`, if they are declared in the global scope, and subsequently redeclared inside a function/block scope, the new value is only limited to the scope it is assigned, and does not affect the value of the globally scoped variable, eg. 
```
let b = 1;
function f() {
  let b = 3;
  console.log(b); // prints 3
}
console.log(b); // prints 1
```

Again, my understanding of Javascript is very limited and it's the first time I encountered many concepts (like `let`, `const`, promise, stream etc.) when doing the task. Please correct me if my understanding is not correct. Thanks!

**Pull request closed without merging**